File: Data Structures & Algorithms/sliding-window-maximum/submission-0.py
import heapq

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        answer = [0] * (len(nums) - k + 1)
        q = []
        
        for i in range(k):
            heapq.heappush(q, (-1 * nums[i], i))

        for i in range(len(nums) - k + 1):
            
            heapq.heappush(q, (-1 * nums[i + k - 1], i + k - 1))

            _, index = heapq.heappop(q)
            while index < i or index >= i + k:
                _, index = heapq.heappop(q)
            
            answer[i] = nums[index]
            heapq.heappush(q, (-1 * nums[index], index))
        return answer

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        if (nums.length == 1 || nums.length == 0) {
            return nums;
        }
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];
        for (int index = 0; index < nums.length; index++) {
            int other = nums.length - index - 1;
            prefix[index] = nums[index];
            suffix[other] = nums[other];
            if (index != 0) {
                prefix[index] *= prefix[index - 1];
                suffix[other] *= suffix[other + 1];
            }
        }
        print(prefix);
        print(suffix);
        int[] answer = new int[nums.length];
        for (int index = 0; index < nums.length; index++) {
            answer[index] = 1;
            if (index != 0) {
                answer[index] = prefix[index - 1];
            }
            if (index != nums.length - 1) {
                answer[index] *= suffix[index + 1];
            }
        }
        return answer;
    }

    private void print(int[] r) {
        for (int j : r) {
            System.out.print(j + ",");
        }
        System.out.println();
    }
}  
